woocommerce/woocommerce
# Add hooks to customize the order receipt generation

### Changes proposed in this Pull Request:

Adds a few extensibility points to [the receipt rendering engine]( https://github.com/woocommerce/woocommerce/pull/43502 )to allow customizing the generated receipts. More precisely:

* A `woocommerce_printable_order_receipt_data` filter to customize the data that gets combined with the receipt template in order to produce the actual receipt.
* A `woocommerce_printable_order_receipt_line_item_display_data` filter to customize the rendering of each order line item. `wp_kses_post` is applied to the HTML returned by this filter for security.
* A `woocommerce_printable_order_receipt_css` filter to customize the CSS styles embedded in the generated receipt.

See the documentation blocks for each hook in [the ReceiptRenderingEngine class](https://github.com/woocommerce/woocommerce/blob/add_hooks_for_receipt_rendering/plugins/woocommerce/src/Internal/ReceiptRendering/ReceiptRenderingEngine.php) for details.

Also in order to support these changes:

- Receipt data is now accessed through a `$data` array in the template (data is not extracted to variables anymore).
- CSS styles are moved to a separate template file.
- All the relevant HTML elements in the receipt get an explicit `id` attribute (helpful when customizing the CSS).

Closes #46122.



### How to test the changes in this Pull Request:

1. Create an order with at least one product.

2. Create a receipt for the order via code or via REST API (see the testing instructions of [the receipts rendering engine pull request]( https://github.com/woocommerce/woocommerce/pull/43502 ) for details), TLDR using `curl`: 

```
curl -X POST -u <user>:<password> http://localhost/wp-json/wc/v3/orders/<order id>/receipt?force_new=true
```

3. Open the URL for the generated receipt in your browser, verify that it renders as expected.

4. Now add the following code snippet at the end of `woocommerce.php` or using [the Code Snippets plugin](https://wordpress.org/plugins/code-snippets/):

```php
add_filter( 'woocommerce_printable_order_receipt_data', function( $data, $order ) {
	$data['texts']['amount_paid_section_title'] = 'How much??';
	$data['notes'][] = 'Order id: ' . $order->get_id() . ' - note added via filter.';
	$data['payment_info'] = [
		'payment_method' => 'woocommerce_payments',
		'card_icon'    => base64_encode( file_get_contents( WC_ABSPATH . "src/Internal/ReceiptRendering/CardIcons/diners.svg" ) ),
		'card_last4'   => '1234',
		'app_name'     => 'some_app_name',
		'aid'          => 'some_file_name',
		'account_type' => 'some_account_type'
	];
	return $data;
}, 10, 2 );

add_filter( 'woocommerce_printable_order_receipt_line_item_display_data', function( $line_item_display_data, $line_item_data, $order ) {
    if( $line_item_data['type'] === 'product' ) {
        $product_id = $line_item_data['item']->get_product()->get_id();
        $order_id = $order->get_id();
        $line_item_display_data['inner_html'] =
            //To test the wp_kses_post, try this instead:
	        //"<script>alert('YOU\'VE BEEN H4CK3D');</script>";
            "<td>${line_item_display_data['row_index']}: ($product_id in $order_id) " . substr( $line_item_display_data['inner_html'], 4 );
    }
    else {
        $line_item_display_data['tr_attributes']['class'] = 'non_product';
    }
    return $line_item_display_data;
}, 10, 3 );

add_filter( 'woocommerce_printable_order_receipt_css', function( $css, $order ) {
    $amount = $order->get_total();
    if($amount < 100) {
        $color = '00F000';
    }
    else if($amount < 1000) {
        $color = 'F0F000';
    }
    else {
        $color = 'F00000';
    }

    return $css . "table#line_items { background-color:#$color; } tr.non_product {font-style: italic}";
}, 10, 2 );
```

5. Regenerate the order receipt (don't forget to add the `force_true` argument to the request).

6. Open the new receipt and verify that it has changed as follows:

* The "Amount paid" section title is now "How much??"
* The background color of the line items table is now green if the total amount of the order is less than 100, yellow if the amount is between 100 and 1000, and green if it's more than 1000.
* Line items corresponding to products have now a prefix that includes the row number, the product id and the order id.
* Line items that don't correspond to products are rendered in italics.
* There's one fake order note.
* There's fake payment information.

### Changelog entry




<details>

#### Significance




#### Type




#### Message 

#### Comment 

</details>
